src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,257 @@
+            private long _pingSentTimestamp;
+            private long _pingCounter;
+            private int _initialBurst;
+            private volatile bool _terminating;
+            private long _minRtt;
+
+            public TimeSpan MinRtt => new TimeSpan(_minRtt);
+
+            public RttEstimator(Http2Connection connection)
+            {
+                _connection = connection;
+                _state = connection._pool.Settings._disableDynamicHttp2WindowSizing ? State.Disabled : State.Init;
+                _pingCounter = 0;
+                _initialBurst = 4;
+                _pingSentTimestamp = default;
+                _minRtt = 0;
+                _terminating = false;
+            }
+
+            internal void OnInitialSettingsSent()
+            {
+                if (_state == State.Disabled) return;
+                _pingSentTimestamp = Stopwatch.GetTimestamp();
+            }
+
+            internal void OnInitialSettingsAckReceived()
+            {
+                if (_state == State.Disabled) return;
+                RefreshRtt();
+                _state = State.Waiting;
+            }
+
+            internal void OnDataOrHeadersReceived()
+            {
+                if (_state != State.Waiting) return;
+
+                if (_terminating)
+                {
+                    _state = State.Disabled;
+                    return;
+                }
+
+                long now = Stopwatch.GetTimestamp();
+                bool initial = _initialBurst > 0;
+                if (initial || now - _pingSentTimestamp > PingIntervalInTicks)
+                {
+                    if (initial) _initialBurst--;
+
+                    // Send a PING
+                    _pingCounter--;
+                    _connection.LogExceptions(_connection.SendPingAsync(_pingCounter, isAck: false));
+                    _pingSentTimestamp = now;
+                    _state = State.PingSent;
+                }
+            }
+
+            internal void OnPingAckReceived(long payload)
+            {
+                if (_state != State.PingSent)
+                {
+                    ThrowProtocolError();
+                }
+                Debug.Assert(payload < 0);
+
+                if (_pingCounter != payload)
+                    ThrowProtocolError();
+
+                RefreshRtt();
+                _state = State.Waiting;
+            }
+
+            internal void OnGoAwayReceived()
+            {
+                _state = State.Disabled;
+            }
+
+            internal void MaximumWindowReached()
+            {
+                // This method is being called from a thread other than the rest of the methods, so do not mess with _state.
+                // OnDataOrHeadersReceived will eventually flip it to State.Disabled.
+                _terminating = true;
+            }
+
+            private void RefreshRtt()
+            {
+                long elapsedTicks = Stopwatch.GetTimestamp() - _pingSentTimestamp;
```

Comment:
Seems like we should assert here about the `_pingSentTimeStamp` value, that it's not `default`

Reply:
And set it back to `default` after we use it, so the assert can check it again

Reply:
Actually I guess you can't do that because you need it later, to manage ping frequency.

I guess you don't really need an assert here because you are careful to check in the callers.